### PR TITLE
Suppress instrumentation based on suppress Context key

### PR DESCRIPTION
**Pull request by LikeTheSalad**
An approach to suppress automatic instrumentations based on the [suppress_instrumentation](https://github.com/open-telemetry/opentelemetry-java/pull/5918) Context key.

### CODE REVIEW

**Comment by LikeTheSalad** [comment]
Hi @jack-berg - If I understood correctly, these should be the changes we were talking about in last week's SIG. We will need to use a Context key defined in the OTel Java repo though, so I guess these changes would have to wait for a new release there that contains [these changes](https://github.com/open-telemetry/opentelemetry-java/pull/5918). Please let me know if there's anything we should improve here or in the core repo's PR!

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/SpanSuppressors.java [hunk 1]
```diff
@@ -85,4 +87,28 @@ public boolean shouldSuppress(Context parentContext, SpanKind spanKind) {
       return true;
     }
   }
+
+  static class ByContextKey implements SpanSuppressor {
+    private final SpanSuppressor delegate;
+    static final ContextKey<Boolean> SUPPRESS_INSTRUMENTATION =
+        ContextKey.named(
+            "suppress_instrumentation"); // todo must use the key defined in the Java SDK.
```
Comment (LikeTheSalad) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/SpanSuppressors.java:95]: I was planning to use the tool recently added to the core project, however, I just realized that my changes are part of the `opentelemetry-exporter-common` lib which this project currently doesn't have as part of its dependencies. So I'm wondering if it's ok to add said dependency here or if it's better to move the context suppressing utility over to the `opentelemetry-api` core subproject?

Comment (trask) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/SpanSuppressors.java:95]: oh no, that's a good point

we could access it via reflection for now, but this seems like a good reason to move the utility over to `opentelemetry-api` in the long run

Comment (LikeTheSalad) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/SpanSuppressors.java]: Got it 👍 I've updated it to use reflection for now.

**Comment by laurit** [comment]
@LikeTheSalad latest dep tests can start failing when there is a new version of tested library. They are not required for merging put you could get them passing by rebasing your pr.

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/SpanSuppressors.java [hunk 2]
```diff
@@ -85,4 +87,48 @@ public boolean shouldSuppress(Context parentContext, SpanKind spanKind) {
...
     }
   }
+
+  static class ByContextKey implements SpanSuppressor {
+    private final SpanSuppressor delegate;
+    private final Method shouldSuppressInstrumentation;
+
+    ByContextKey(SpanSuppressor delegate) {
+      this.delegate = delegate;
+      Method shouldSuppressInstrumentation;
```
Comment (mateuszrzeszutek) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/SpanSuppressors.java:97]: Too bad Android desugaring doesn't support `MethodHandle`s 😢

**Review by mateuszrzeszutek — APPROVED** [review]

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/SpanSuppressors.java (hunk 2 above)
Comment (LikeTheSalad) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/SpanSuppressors.java:97]: Ikr 😩

**Review by laurit — APPROVED** [review]

**Pull request merged**